app/controllers/IsThisYourBusinessController.scala

```diff
@@ -105,29 +90,54 @@ class IsThisYourBusinessController @Inject() (
   }
 
   def onPageLoad(mode: Mode, regime: Regime): Action[AnyContent] = (identify(regime) andThen getData.apply andThen requireData(regime)).async {
-    implicit request => result(mode, regime, form)
+    implicit request =>
+      buildRegistrationRequest match {
+        case Some(registrationRequest) =>
+          matchingService.sendBusinessRegistrationInformation(regime, registrationRequest).flatMap {
+            case Right(response) =>
+              request.userAnswers.set(RegistrationInfoPage, response).map(sessionRepository.set)
+              result(mode, regime, form, response)
+            case _ =>
+              Future.successful(Redirect(routes.NoRecordsMatchedController.onPageLoad(regime)))
+          }
+        case _ =>
+          renderer
+            .render("thereIsAProblem.njk", Json.obj("regime" -> regime.toUpperCase, "emailAddress" -> appConfig.emailEnquiries))
+            .map(ServiceUnavailable(_))
+      }
   }
 
-  private def matchBusinessInfo(regime: Regime)(implicit request: DataRequest[AnyContent]): Future[Either[ApiError, RegistrationInfo]] =
-    (for {
-      utr          <- getEither(UTRPage)
-      businessName <- getEither(BusinessNamePage).orElse(getEither(SoleNamePage).map(_.fullName))
-      businessType <- getEither(BusinessTypePage)
-      dob = request.userAnswers.get(SoleDateOfBirthPage)
-      registrationInfo <- EitherT(matchingService.sendBusinessRegistrationInformation(regime, RegistrationInfo.build(businessType, businessName, utr, dob)))
-    } yield registrationInfo).value
-
   def onSubmit(mode: Mode, regime: Regime): Action[AnyContent] = (identify(regime) andThen getData.apply andThen requireData(regime)).async {
     implicit request =>
+      val thereIsAProblem =
+        renderer.render("thereIsAProblem.njk", Json.obj("regime" -> regime.toUpperCase, "emailAddress" -> appConfig.emailEnquiries)).map(ServiceUnavailable(_))
       form
         .bindFromRequest()
         .fold(
-          formWithErrors => result(mode, regime, formWithErrors),
+          formWithErrors =>
+            request.userAnswers.get(RegistrationInfoPage).fold(thereIsAProblem) {
+              registrationInfo =>
+                result(mode, regime, formWithErrors, registrationInfo)
```

Comment:
I think this should call the render method rather than result. I don't think we need to make the call to get the subscription ID or create the enrolment just to render the form with errors:

`                val name     = registrationInfo.name.getOrElse("")`
`                val address  = registrationInfo.address.getOrElse(AddressResponse("", None, None, None, None, ""))`
`                render(mode, regime, formWithErrors, name, address).map(BadRequest(_))`

Reply:
Good point!